### PR TITLE
fix iterators may not be closed

### PR DESCRIPTION
**Pull request by javeme**
Change-Id: I3709d96fd2114fa782d6f28f8853b32d65fbd22b

### CODE REVIEW

File: src/main/java/com/baidu/hugegraph/iterator/FlatMapperIterator.java [hunk 1]
```diff
@@ -29,48 +29,63 @@
...
                 this.current = result;
                 return true;
             }
         }
-        this.results = null;
+        this.resetResults();
         return false;
     }
+
+    protected final void resetResults() {
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/iterator/FlatMapperIterator.java]: resetBatchIterator

File: src/test/java/com/baidu/hugegraph/unit/iterator/BatchMapperIteratorTest.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+    @Test
+    public void testNext() {
+        Iterator<Integer> results;
+
+        results = new BatchMapperIterator<>(1, DATA1.iterator(), MAPPER);
+        // Call next() without hasNext()
+        Assert.assertEquals(1, results.next());
+
+        results = new BatchMapperIterator<>(1, DATA3.iterator(), MAPPER);
```
Comment (Linary) [inline src/test/java/com/baidu/hugegraph/unit/iterator/BatchMapperIteratorTest.java]: add a test with batch size 2

File: src/test/java/com/baidu/hugegraph/unit/iterator/BatchMapperIteratorTest.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            return batch.iterator();
+        });
+        Assert.assertEquals(4, results.next());
+        Assert.assertEquals(5, results.next());
+        Assert.assertEquals(6, results.next());
+        Assert.assertFalse(results.hasNext());
+    }
+
+    @Test
+    public void testMapperTHenReturn2X() {
```
Comment (Linary) [inline src/test/java/com/baidu/hugegraph/unit/iterator/BatchMapperIteratorTest.java]: THen -> Then

File: src/main/java/com/baidu/hugegraph/iterator/BatchMapperIterator.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                assert this.current == none();
+                this.current = result;
+                return true;
+            }
+        }
+        this.resetResults();
+        return false;
+    }
+
+    protected final void resetResults() {
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/iterator/BatchMapperIterator.java]: resetBatchIterator

**Pull request merged**